src/common/client.go

```diff
@@ -16,38 +16,75 @@ func ReconcileService(client *opslevel.Client, service ServiceRegistration) {
+	case serviceAliasesResult_NoAliasesMatched:
 		newService, newServiceErr := createService(client, service)
 		if newServiceErr != nil {
+			log.Warn().Msgf("[%s] api error during service creation ... skipping reconciliation.\n\tREASON: %v", service.Name, newServiceErr)
 			return
 		}
 		foundService = newService
-	}
-	if needsUpdate {
+	case serviceAliasesResult_AliasMatched:
 		updateService(client, service, foundService)
+
+	case serviceAliasesResult_MultipleServicesFound:
+		log.Warn().Msgf("[%s] found multiple services with aliases = [\"%s\"].  cannot know which service to target for update ... skipping reconciliation", service.Name, strings.Join(service.Aliases, "\", \""))
+		return
+	case serviceAliasesResult_APIErrorHappened:
+		log.Warn().Msgf("[%s] api error during service lookup by alias.  unable to guarentee service was found or not ... skipping reconciliation", service.Name)
+		return
 	}
+
 	handleAliases(client, service, foundService)
 	handleTags(client, service, foundService)
 	handleTools(client, service, foundService)
 	handleRepositories(client, service, foundService)
 	log.Info().Msgf("[%s] Finished processing data", foundService.Name)
 }
 
-func findService(client *opslevel.Client, registration ServiceRegistration) (*opslevel.Service, bool) {
+type serviceAliasesResult string
+
+const (
+	serviceAliasesResult_NoAliasesMatched      serviceAliasesResult = "NoAliasesMatched"
+	serviceAliasesResult_AliasMatched          serviceAliasesResult = "AliasMatched"
+	serviceAliasesResult_MultipleServicesFound serviceAliasesResult = "MultipleServicesFound"
+	serviceAliasesResult_APIErrorHappened      serviceAliasesResult = "APIErrorHappened"
+)
+
+// This function has 4 outcomes that can happen while looping over the aliases list
+// serviceAliasesResult_NoAliasesMatched - means that all API calls succeeded and none of the aliases matched an existing service
+// serviceAliasesResult_AliasMatched - means that all the API calls succeeded and a single service was found matching 1 of N aliases
+// serviceAliasesResult_MultipleServicesFound - means that all API calls succeeded but multiple services were returning means the list of aliases does not definitively describe a single service and might be a configuration problem
+// serviceAliasesResult_APIErrorHappened - means that 1 of N aliases got an 4xx/5xx and thereforce we cannot say 100% that the services doesn't exist
+func validateServiceAliases(client *opslevel.Client, registration ServiceRegistration) (*opslevel.Service, serviceAliasesResult) {
+	var gotError error
+	foundServices := map[string]*opslevel.Service{}
 	for _, alias := range registration.Aliases {
 		foundService, err := client.GetServiceWithAlias(alias)
-		if err == nil && foundService.Id != nil {
-			log.Info().Msgf("[%s] Reconciling service found with alias '%s' ...", foundService.Name, alias)
-			return foundService, true
+		if err != nil {
+			gotError = err
+		} else {
+			if foundService.Id != nil {
```

Comment:
Would there be a case where `foundService.Id == nil`?

Reply:
yes that happens when the API call `GetServiceWithAlias` is 2xx but there was no service with that "alias"

The API actually responds 2xx when a service with the requested aliases is not an alias on any server and because golang is strongly typed the service object is initialized with default values - in this case ID == nil

so when as err == nil and foundService.Id == nil that means alias is not a valid alias on any service.

Reply:
It should also be noted this function is doing 2 things:

- trying to find all the services that match any of the aliases
- trying to protect against API failures 4xx/5xx to let the caller know - hey this function didn't work as expected - hence change from output being a bool to output being an enum

Reply:
Okay that makes sense, I was unsure why we would have a `foundService` object if `GetServiceWithAlias` found nothing but knowing now that its initialized with default values, it makes sense 👍 

LGTM